cmd/argo/commands/get.go

```diff
@@ -116,25 +118,301 @@ func printWorkflowHelper(wf *wfv1.Workflow) {
+	}
+	return false
+}
+
+func convertToRenderTrees(wf *wfv1.Workflow) map[string]renderNode {
+
+	renderTreeRoots := make(map[string]renderNode)
+
+	// Used to store all boundary nodes so future render children can attach
+	// Maps node Name -> *boundaryNode
+	boundaryNodeMap := make(map[string]*boundaryNode)
+	// Used to store children of a boundary node that has not been parsed yet
+	// Maps boundary Node name -> array of render Children
+	parentBoundaryMap := make(map[string][]renderNode)
+
+	// Used to store Non Boundary Parent nodes so render children can attach
+	// Maps non Boundary Parent Node name -> *nonBoundaryParentNode
+	nonBoundaryParentMap := make(map[string]*nonBoundaryParentNode)
+	// Used to store children which have a Non Boundary Parent from rendering perspective
+	// Maps non Boundary render Children name -> *nonBoundaryParentNode
+	nonBoundaryParentChildrenMap := make(map[string]*nonBoundaryParentNode)
+
+	// We have to do a 2 pass approach because anything that is a child
+	// of a nonBoundaryParent and also has a boundaryID we may not know which
+	// parent to attach to if we didn't see the nonBoundaryParent earlier
+	// in a 1 pass strategy
+
+	// 1st Pass Process enough of nonBoundaryParent nodes to know all their children
+	for id, status := range wf.Status.Nodes {
+		if isNonBoundaryParentNode(status.Type) {
+			n := nonBoundaryParentNode{node: id}
+			nonBoundaryParentMap[id] = &n
+
+			for _, child := range status.Children {
+				nonBoundaryParentChildrenMap[child] = &n
+			}
+		}
+	}
+
+	// 2nd Pass process everything
+	for id, status := range wf.Status.Nodes {
+		switch {
+		case isBoundaryNode(status.Type):
+			n := boundaryNode{node: id}
+			boundaryNodeMap[id] = &n
+			// Attach to my parent if needed
+			if attachToParent(wf, &n, nonBoundaryParentChildrenMap,
+				status.BoundaryID, boundaryNodeMap, parentBoundaryMap) {
+				renderTreeRoots[n.getID()] = &n
+			}
+			// Attach nodes who are in my boundary already seen before me to me
+			for _, val := range parentBoundaryMap[id] {
+				n.boundaryContained = insertSorted(wf, n.boundaryContained, val)
+			}
+		case isNonBoundaryParentNode(status.Type):
+			nPtr, ok := nonBoundaryParentMap[id]
+			if !ok {
+				log.Fatal("Unable to lookup node " + id)
+				return nil
+			}
+			// Attach to my parent if needed
+			if attachToParent(wf, nPtr, nonBoundaryParentChildrenMap,
+				status.BoundaryID, boundaryNodeMap, parentBoundaryMap) {
+				renderTreeRoots[nPtr.getID()] = nPtr
+			}
+			// All children attach directly to the nonBoundaryParents since they are already created
+			// in pass 1 so no need to do that here
+		case isExecutionNode(status.Type):
+			n := executionNode{node: id}
+			// Attach to my parent if needed
+			if attachToParent(wf, &n, nonBoundaryParentChildrenMap,
+				status.BoundaryID, boundaryNodeMap, parentBoundaryMap) {
+				renderTreeRoots[n.getID()] = &n
+			}
+			// Execution nodes don't have children
```

Comment:
This comment isn't correct. In the DAG case (e.g. dag diamond), pods have children:

```
    dag-diamond-qg6dr-3228393458:
      boundaryID: dag-diamond-qg6dr
      children:
      - dag-diamond-qg6dr-3211615839
      - dag-diamond-qg6dr-3194838220
      finishedAt: 2018-02-04T02:19:07Z
      id: dag-diamond-qg6dr-3228393458
      name: dag-diamond-qg6dr.A
      phase: Succeeded
      startedAt: 2018-02-04T02:18:59Z
      type: Pod
```